### PR TITLE
Add custom attribute decoder to S.R.Metadata vNext

### PR DESCRIPTION
**Pull request by nguerrera**
Port from dev/metadata branch.

It is internalized out of the upcoming release as with the metadata writer and signature decoder, and equally subject to API review.

@tmat

### CODE REVIEW

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+using System.Collections.Immutable;
+
+namespace System.Reflection.Metadata.Decoding
+{
+    /// <summary>
+    /// Decodes custom attribute blobs.
+    /// </summary>
+    internal struct CustomAttributeDecoder<TType>
+    {
+        private ICustomAttributeTypeProvider<TType> _provider;
```
Comment (stephentoub) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs]: readonly?

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+        }
+
+        private CustomAttributeTypedArgument<TType> DecodeArgument(ref BlobReader valueReader, ArgumentTypeInfo info)
+        {
+            if (info.TypeCode == SerializationTypeCode.TaggedObject)
+            {
+                info = DecodeNamedArgumentType(ref valueReader);
+            }
+
+            // PERF_TODO: Cache/reuse common arguments to avoid boxing (small integers, true, false).
```
Comment (stephentoub) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs]: Is this tracked somewhere?

**Comment by stephentoub** [comment]
LGTM

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+        }
+
+        private ImmutableArray<CustomAttributeTypedArgument<TType>> DecodeFixedArguments(ref BlobReader signatureReader, ref BlobReader valueReader, int count)
+        {
+            if (count == 0)
+            {
+                return ImmutableArray<CustomAttributeTypedArgument<TType>>.Empty;
+            }
+
+            var arguments = new CustomAttributeTypedArgument<TType>[count];
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs]: Note to self: use builder + MoveToImmutable()  (everywhere we do new T[count] + ImmutableArray.Create().

Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs]: Perhaps we also want to use pooled builder (we can copy it from Roslyn if we already don't).

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs [hunk 4]
```diff
@@ -0,0 +1,373 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs]: Out of date copyright header.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs (hunk 1 above)
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs]: Fixed.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs (hunk 3 above)
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs]: Maybe, there's a tradeoff between allocating small outer builder object vs. copying from pooled builder to final destination, plus any waste associated with unused pooled objects. I'm not sure which is better. For now, I went with same pattern as signature decoder (MoveToImmutable). If we should switch to pooled buildes, we should do it across the board.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs (hunk 2 above)
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/CustomAttributeDecoder.cs]: It is now: https://github.com/dotnet/corefx/issues/6533 :)

**Comment by nguerrera** [comment]
@dotnet-bot test this please (it's been running for a very long time and there are no Jenkins detail links)

**Pull request merged**